sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/TrafficController.scala

```diff
@@ -0,0 +1,142 @@
+  override def call(): T = callable.call()
+}
+
+/**
+ * Throttle interface to be implemented by different throttling strategies.
+ *
+ * Currently, only HostMemoryThrottle is implemented, which limits the maximum in-flight host
+ * memory bytes. In the future, we can add more throttling strategies, such as limiting the
+ * device memory usage, the number of tasks, etc.
+ */
+trait Throttle {
+
+  /**
+   * Returns true if the task can be accepted, false otherwise.
+   * TrafficController will block the task from being scheduled until this method returns true.
+   */
+  def canAccept[T](task: Task[T]): Boolean
+
+  /**
+   * Callback to be called when a task is scheduled.
+   */
+  def taskScheduled[T](task: Task[T]): Unit
+
+  /**
+   * Callback to be called when a task is completed, either successfully or with an exception.
+   */
+  def taskCompleted[T](task: Task[T]): Unit
+}
+
+/**
+ * Throttle implementation that limits the total host memory used by the in-flight tasks.
+ */
+class HostMemoryThrottle(val maxInFlightHostMemoryBytes: Long) extends Throttle {
+  private var totalHostMemoryBytes: Long = 0
+
+  override def canAccept[T](task: Task[T]): Boolean = {
+    totalHostMemoryBytes + task.hostMemoryBytes <= maxInFlightHostMemoryBytes
+  }
+
+  override def taskScheduled[T](task: Task[T]): Unit = {
+    totalHostMemoryBytes += task.hostMemoryBytes
+  }
+
+  override def taskCompleted[T](task: Task[T]): Unit = {
+    totalHostMemoryBytes -= task.hostMemoryBytes
+  }
+
+  def getTotalHostMemoryBytes: Long = totalHostMemoryBytes
+}
+
+/**
+ * TrafficController is responsible for blocking tasks from being scheduled when the throttle
+ * is exceeded. It also keeps track of the number of tasks that are currently scheduled.
+ *
+ * This class is thread-safe as it is used by multiple tasks.
+ */
+class TrafficController protected[rapids] (throttle: Throttle) {
+
+  @GuardedBy("this")
+  private var numTasks: Int = 0
+
+  /**
+   * Blocks the task from being scheduled until the throttle allows it. If there is no task
+   * currently scheduled, the task is scheduled immediately even if the throttle is exceeded.
+   */
+  def blockUntilRunnable[T](task: Task[T]): Unit = synchronized {
+    if (numTasks > 0) {
+      while (!throttle.canAccept(task)) {
+        wait(100)
+      }
+    }
+    numTasks += 1
+    throttle.taskScheduled(task)
+  }
+
+  def taskCompleted[T](task: Task[T]): Unit = synchronized {
+    numTasks -= 1
+    throttle.taskCompleted(task)
+    notifyAll()
+  }
+
+  def numScheduledTasks: Int = synchronized {
+    numTasks
+  }
+}
+
+object TrafficController {
+
+  private var instance: TrafficController = _
+
+  /**
+   * Initializes the TrafficController singleton instance.
+   * This is called once per executor.
+   */
+  def initialize(conf: RapidsConf): Unit = synchronized {
+    if (conf.isAsyncOutputWriteEnabled && instance == null) {
+      instance = new TrafficController(
```

Comment:
it might be nice to add a log message for the executor here "AsyncOutputWrite enabled". We may need one for each "async" thing, "AsyncShuffleWrite enabled"?

Reply:
It's currently logged in [`ColumnarOutputWriter`](https://github.com/NVIDIA/spark-rapids/pull/11730/files/31d7d3017b070a94bc206594a4981604139c5710#diff-2a72b3217aa1da8bb1f2cfb2c9674de901dc27232580bb1f72182d6b3805bd05R94). Do you think it's better to do it here instead?

Reply:
I would make it an info level thing, so that we know that it was on when we get a log to debug.